### PR TITLE
adds onadd & onremove methods to componentdependencies

### PR DESCRIPTION
**Pull request by PaulRitter**

### CODE REVIEW

File: Robust.Shared/GameObjects/ComponentDependencies/ComponentDependencyManager.cs [hunk 1]
```diff
@@ -112,38 +126,73 @@ private void SetField(object o, int offset, object? value)
...
                 }
 
                 var offset = GetFieldOffset(objType, field);
 
-                queries[i] = (field.FieldType, offset);
+                var attribute = attributeFields[i].attribute;
+
+                if (attribute == null)
+                {
+                    throw new NotImplementedException();//todo check if attribute null
```
Comment (PaulRitter) [inline Robust.Shared/GameObjects/ComponentDependencies/ComponentDependencyManager.cs]: @PJB3005 should i really check if it failed to get the attribute? it only runs the select (l139) on fields where `Attribute.IsDefined(f, typeof(ComponentDependencyAttribute)`

**Comment by PaulRitter** [comment]
Ready for review.
Idea behind this is to have a place to sanely (un)subscribe to component events etc

File: Robust.Shared/GameObjects/ComponentDependencies/ComponentDependencyManager.cs [hunk 2]
```diff
@@ -20,7 +20,11 @@ public class ComponentDependencyManager : IComponentDependencyManager
         /// <summary>
         /// Cache of queries and their corresponding field offsets
         /// </summary>
-        private readonly Dictionary<Type, (Type Query, int FieldMemoryOffset)[]> _componentDependencyQueries = new Dictionary<Type, (Type Query, int FieldMemoryOffset)[]>();
+        private readonly
+            Dictionary<Type, (Type Query, int FieldMemoryOffset, MethodInfo? OnAddMethod, MethodInfo? OnRemoveMethod)[]>
```
Comment (PJB3005) [inline Robust.Shared/GameObjects/ComponentDependencies/ComponentDependencyManager.cs]: 1. At this point these tuples are so large you should switch them to a full blown class.
2. Caching `MethodInfo` instances and calling `.Invoke()` on them is like 90x slower than caching a delegate and calling it. See https://codeblog.jonskeet.uk/2008/08/09/making-reflection-fly-and-exploring-delegates/

**Review by PJB3005 — CHANGES_REQUESTED** [review]

**Comment by PaulRitter** [comment]
@PJB3005 ready for review

File: Robust.Shared/GameObjects/ComponentDependencies/ComponentDependencyManager.cs [hunk 3]
```diff
@@ -112,44 +125,83 @@ private void SetField(object o, int offset, object? value)
...
-                queries[i] = (field.FieldType, offset);
+                var attribute = attributeFields[i].attribute!;
+
+                var methods = objType.GetRuntimeMethods().ToArray();
+
+                Action<object>? onAddMethod = null;
+                MethodInfo? getterMethod = null;
+                if (attribute.OnAddMethodName != null)
+                {
+                    getterMethod = typeof(ComponentDependencyManager).GetMethod("GetEventMethodDelegate",
```
Comment (PJB3005) [inline Robust.Shared/GameObjects/ComponentDependencies/ComponentDependencyManager.cs]: Move this method get to outside the two if statements so you don't duplicate it. If you care about not bothering with it when no add/remove methods are specified just wrap the entire thing in a check for either not being null.

File: Robust.Shared/GameObjects/ComponentDependencies/ComponentDependencyManager.cs [hunk 4]
```diff
@@ -202,13 +270,37 @@ public ComponentDependencyValueTypeException(Type componentType, FieldInfo field
...
+        public readonly FieldInfo Field;
 
         public ComponentDependencyNotNullableException(Type componentType, FieldInfo field) : base($"Field {field} of Type {componentType} is marked as ComponentDependency, but does not have ?(Nullable)-Flag!")
         {
             ComponentType = componentType;
             Field = field;
         }
     }
+
+    public abstract class ComponentDependencyInvalidMethodName : Exception
```
Comment (PJB3005) [inline Robust.Shared/GameObjects/ComponentDependencies/ComponentDependencyManager.cs]: This name needs to end with `Exception`.

File: Robust.UnitTesting/Shared/GameObjects/ComponentDependencies_Tests.cs [hunk 5]
```diff
@@ -131,6 +143,13 @@ private class TestSixComponent : Component
             [ComponentDependency] public TestFiveComponent thing = null!;
         }
 
+        private class TestSevenComponent : Component
+        {
+            public override string Name => "TestSeven";
+
+            [ComponentDependency("ABCDEF")] public TestFiveComponent? thing;
```
Comment (PJB3005) [inline Robust.UnitTesting/Shared/GameObjects/ComponentDependencies_Tests.cs]: Mute these compiler warnings.

**Review by PJB3005 — CHANGES_REQUESTED** [review]

**Comment by PaulRitter** [comment]
reviews adressed

**Review by PJB3005 — APPROVED** [review]

**Pull request merged**